src/redis-cli.c

```diff
@@ -418,92 +425,273 @@ static sds cliVersion(void) {
+                    multiple = 1;
+                }
+                else if (!strcmp(flag, "multiple_token")) {
+                    multipleToken = 1;
+                }
+            }
+        }
+    }
+
+    /* Then build the "repeating part" of the argument string. */
+    if (!strcmp(type, "key") ||
+        !strcmp(type, "string") ||
+        !strcmp(type, "integer") ||
+        !strcmp(type, "double") ||
+        !strcmp(type, "pattern") ||
+        !strcmp(type, "unix-time") ||
+        !strcmp(type, "token"))
+    {
+        repeatPart = sdscat_orempty(repeatPart, name);
+    } else if (!strcmp(type, "oneof")) {
+        repeatPart = cliConcatArguments(repeatPart, arguments, "|");
+    } else if (!strcmp(type, "block")) {
+        repeatPart = cliConcatArguments(repeatPart, arguments, " ");
+    } else if (strcmp(type, "pure-token") != 0) {
+        fprintf(stderr, "Unknown type '%s' set for argument '%s'\n", type, name);
+    }
+
+    /* Finally, build the parameter string. */
+    if (tokenPart[0] != '\0' && strcmp(type, "pure-token") != 0) {
+        tokenPart = sdscat(tokenPart, " ");
+    }
+    if (optional) {
+        params = sdscat(params, "[");
+    }
+    params = sdscat(params, tokenPart);
+    params = sdscat(params, repeatPart);
+    if (multiple) {
+        params = sdscat(params, " [");
+        if (multipleToken) {
+            params = sdscat(params, tokenPart);
+        }
+        params = sdscat(params, repeatPart);
+        params = sdscat(params, " ...]");
+    }
+    if (optional) {
+        params = sdscat(params, "]");
+    }
+    sdsfree(tokenPart);
+    sdsfree(repeatPart);
+    return params;
+}
+
+static void cliCreateHelpEntry(helpEntry *help, char *cmdname, char *subcommandname) {
+    help->argc = subcommandname ? 2 : 1;
+    help->argv = zmalloc(sizeof(sds) * help->argc);
+    help->argv[0] = sdsnew(cmdname);
+    sdstoupper(help->argv[0]);
+    if (subcommandname) {
+        /* Subcommand name is two words separated by a pipe character. */
+        help->argv[1] = sdsnew(strchr(subcommandname, '|') + 1);
+        sdstoupper(help->argv[1]);
+    }
+    sds fullname = sdsnew(help->argv[0]);
+    if (subcommandname) {
+        fullname = sdscat(fullname, " ");
+        fullname = sdscat(fullname, help->argv[1]);
+    }
+    help->full = fullname;
+    help->type = CLI_HELP_COMMAND;
+
+    struct commandHelp *ch = zmalloc(sizeof(*ch));
+    help->org = ch;
+    ch->name = help->full;
+    ch->params = sdsempty();
+}
+
+/* Initialize a command help entry for the command/subcommand described in specs.
+ * Returns the index of the next available position in the help entries table. */
+static helpEntry *cliInitCommandHelpEntry(char *cmdname, char *subcommandname, helpEntry *next, redisReply *specs) {
+    helpEntry *help = next++;
+    cliCreateHelpEntry(help, cmdname, subcommandname);
+
+    assert(specs->type == (config.resp3 ? REDIS_REPLY_MAP : REDIS_REPLY_ARRAY));
```

Comment:
Would you get rid of all the asserts? I just put them in for consistency. There are element type assertions elsewhere in the file. Or I could change this assert to accept either type. The actual handling code is the same either way.

If the reply data is wrong for some reason, the code will crash somewhere. If it doesn't crash on an assertion it will crash somewhere else. I don't have a strong attachment to the asserts, though. Whatever makes sense to you.

Reply:
i'm ok with the assert, i just don't think we should depend on `config.resp3` (in asserts or `if`s`, but rather just handle what we can, and fail on what we can't handle)

Reply:
Done.